src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java

```diff
@@ -15,55 +14,94 @@
 import org.apache.parquet.schema.Type;
 
-/** Helper class to find all leaf sub-columns in an immutable schema given a dot path. */
+/** Helper class to find all leaf columns in an immutable schema given a fieldId. */
 public class SubColumnFinder {
-  static class SubtreeInterval {
+
+  /**
+   * Helper class to store the start and end index of the interval of leaf columns of a node in the
+   * list and the dot path of the node.
+   */
+  static class SubtreeInfo {
+    /* Start index of the leaf column in the list. */
     final int startTag;
+
+    /* End index of the leaf column in the list. */
     final int endTag;
 
-    SubtreeInterval(int startTag, int endTag) {
+    /* Dot path of the node. */
+    final String dotPath;
+
+    SubtreeInfo(int startTag, int endTag, String dotPath) {
       this.startTag = startTag;
       this.endTag = endTag;
+      this.dotPath = dotPath;
     }
   }
 
+  /* A list to store all leaf columns field id in preorder traversal. */
   private final List<String> list;
-  private final Map<String, SubtreeInterval> accessMap;
+
+  /* A map to cache query result, avoid recursive query during runtime. */
+  private final Map<Type.ID, SubtreeInfo> accessMap;
 
   public SubColumnFinder(MessageType schema) {
     accessMap = new HashMap<>();
     list = new ArrayList<>();
     build(schema, null);
   }
 
-  public List<String> getSubColumns(String dotPath) {
-    if (!accessMap.containsKey(dotPath)) {
-      throw new IllegalArgumentException(String.format("Column %s not found in schema", dotPath));
+  /**
+   * Get all leaf sub-column's field id of a node in the schema.
+   *
+   * @param id Field id of the node
+   * @return List of sub-column's field id
+   */
+  public List<String> getSubColumns(Type.ID id) {
+    if (!accessMap.containsKey(id)) {
+      throw new IllegalArgumentException(String.format("Field %s not found in schema", id));
     }
-    SubtreeInterval interval = accessMap.get(dotPath);
+    SubtreeInfo interval = accessMap.get(id);
     return Collections.unmodifiableList(list.subList(interval.startTag, interval.endTag));
   }
 
-  private void build(Type node, String dotPath) {
-    if (dotPath == null) {
+  /**
+   * Get the dot path of a node in the schema.
+   *
+   * @param id Field ID of the node
+   * @return Dot path of the node
+   */
+  public String getDotPath(Type.ID id) {
+    if (!accessMap.containsKey(id)) {
+      throw new IllegalArgumentException(String.format("Field %s not found in schema", id));
+    }
+    return accessMap.get(id).dotPath;
+  }
+
+  private void build(Type node, List<String> currentPath) {
+    if (currentPath == null) {
       /* Ignore root node type name (bdec or schema) */
-      dotPath = "";
-    } else if (dotPath.isEmpty()) {
-      dotPath = node.getName();
+      currentPath = new ArrayList<>();
     } else {
-      dotPath = concatDotPath(dotPath, node.getName());
+      currentPath.add(node.getName());
```

Comment:
Something feels off here. You're passing a mutable list down every recursion level, and then at the end creating a dotPath out of the whole list of levels. 

Lets say you called build(node, null) on a column like so:
OBJECT col1 (OBJECT col11(ARRAY col111), OBJECT col12(MAP col121))

What will the state of the SubColumnFinder be when col1 is added to SubColumnFinder?